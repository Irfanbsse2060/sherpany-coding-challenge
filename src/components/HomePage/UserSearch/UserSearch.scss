@import "../../../styles/variables";

.user-search {
  display: grid;
  grid-row-gap: 1rem;


  @media #{$extra-small} {
    grid-template-columns: 100%;
  }

  @media #{$medium} {
    grid-template-columns: 50% 50%;
  }

  @media #{$extra-large} {
    grid-template-columns: 25% 75%;
  }

  &__action {
    display: flex;
    align-items: flex-end;
    justify-items: start;
    button {
      margin: 0px 10px;
    }
  }
}
