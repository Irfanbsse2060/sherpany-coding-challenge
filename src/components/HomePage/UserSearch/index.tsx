import React, {ChangeEvent, FormEvent, MouseEvent} from "react";

import './UserSearch.scss'
import {Button, ButtonType, TextField} from '../../UI'
import {UserSearchInterface} from "../types";

interface formProps {
    onSubmit: (userSearch:UserSearchInterface) => void
    onClearFilter: () => void
}

export default function UserSearch({onSubmit, onClearFilter}: formProps) {
    const initialState = {name: ''}
    const [state, setState] = React.useState(initialState)
    const formRef = React.useRef<HTMLFormElement>(null);

    const handleChange = (event: ChangeEvent<HTMLInputElement | HTMLTextAreaElement>) => setState({
        ...state,
        [event.target.name]: event.target.value
    })

    const {name} = state

    const handleSubmit = (event: FormEvent<HTMLFormElement> | MouseEvent<HTMLButtonElement>) => {
        event.preventDefault()
        if (formRef && formRef.current) {
            if (formRef.current.checkValidity()) {
                onSubmit(state)
            } else
                formRef.current.reportValidity();
        }
    }

    return (
        <form ref={formRef} onSubmit={handleSubmit} className='user-search'>
            <TextField label='Name' name='name' value={name} required={true} handleChange={handleChange}/>
            <div className='user-search__action'>
                <Button type={ButtonType.primary} handleClick={handleSubmit} text={'Search'}/>
                <Button type={ButtonType.secondary} handleClick={()=>{
                    setState(initialState)
                    onClearFilter()
                }} text={'Clear search'}/>
            </div>
        </form>
    )


}
