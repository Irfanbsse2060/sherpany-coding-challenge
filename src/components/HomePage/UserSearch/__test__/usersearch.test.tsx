import {render, fireEvent} from '@testing-library/react';
import UserSearch from '../index'
import React from "react";

describe('UserSearch', () => {

    test('should be able to render', () => {
        const {container} = render(<UserSearch onClearFilter={jest.fn} onSubmit={jest.fn()}/>)
        expect(container).toMatchSnapshot()
    })

    test('should not call onSubmit if form is not valid', () => {
        const mockSubmitMethod = jest.fn()
        const {getByTestId} = render(<UserSearch onSubmit={mockSubmitMethod} onClearFilter={jest.fn}/>)
        fireEvent.click(getByTestId('submit-bt'))
        expect(mockSubmitMethod).not.toHaveBeenCalled()
    });


    test('should clear the text when user click on clear button', () => {

        const {getByTestId} = render(<UserSearch onSubmit={jest.fn()} onClearFilter={jest.fn}/>)
        fireEvent.change(getByTestId('text_field_name'), {target: {value: 'test'}})
        expect(getByTestId('text_field_name').value).toEqual('test')

        fireEvent.click(getByTestId('clear-bt'))

        expect(getByTestId('text_field_name').value).toEqual('')
    });

    test('Should called onSubmit method with right form state', () => {

        const name = 'test'
        const mockSubmitMethod = jest.fn()
        const {getByTestId} = render(<UserSearch onSubmit={mockSubmitMethod} onClearFilter={jest.fn}/>)
        fireEvent.change(getByTestId('text_field_name'), {target: {value: name}})

        fireEvent.click(getByTestId('submit-bt'))

        expect(mockSubmitMethod).toHaveBeenCalledTimes(1)
        expect(mockSubmitMethod).toHaveBeenCalledWith({name})

    });


})
