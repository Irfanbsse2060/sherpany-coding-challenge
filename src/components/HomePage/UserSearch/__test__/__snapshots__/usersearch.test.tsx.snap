// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserSearch should be able to render 1`] = `
<div>
  <form
    class="user-search"
  >
    <div
      class="text-field"
    >
      <label
        class="text-field__label"
      >
        Name
      </label>
      <input
        class="text-field__input"
        data-testid="text_field_name"
        name="name"
        required=""
        type="text"
        value=""
      />
    </div>
    <div
      class="user-search__action"
    >
      <button
        class="custom-button custom-button--primary"
        data-testid="submit-bt"
      >
        Search
      </button>
      <button
        class="custom-button custom-button--secondary"
        data-testid="clear-bt"
      >
        Clear search
      </button>
    </div>
  </form>
</div>
`;
